demo04_basic_attacker.py

```diff
@@ -13,17 +13,16 @@
 
 def move(bot, state):
     enemy = bot.enemy
-    # we need to create a dictionary to keep information (state) along rounds
-    # the state object will be passed untouched at every new round
-    if state is None:
-        # initialize the state dictionary
-        state = {}
-        # each bot needs its own state dictionary to keep track of the
-        # food targets
+
+    # The state dictionary is initially empty, so it evaluates as False.
+    if not state:
```

Comment:
I am not sure about using this pattern. While it works well in this case, because we know that `state` is a dict, it is in general a dangerous thing to do, because many things evaluate to `False`, including `0`. And, the students had a problem understanding when the condition `state is None` evaluates to `True`, so I think they would have the same issues with `not state`. What I mean is that it would be clearer to have something like:

```python
if bot.round == 1 and bot.turn == 0:
    ...
```
It is more verbose, but makes it clear that this is initialization code, and doesn't use the side effect of the state dict being empty at the beginning.
What do you think? @keszybz : would you update the PR accordingly if you agree?

Reply:
`If state == {}` ?

Reply:
> `if state == {}` ?

this does not address my main point, i.e. that students where confused abot **when** that code gets executed, so I think it would be better to have the condition to reflect the time instead of the nature of the `state` object.

Reply:
So when testing the `move` function for the second bot, one would have to create a game state manually in advance?

Reply:
> So when testing the `move` function for the second bot, one would have to create a game state manually in advance?

Yes, that reflects what happens in the real game. One of our tests should do that, so it is clear how to do it.

Reply:
I changed this to `state == {}` (before reading this discussion ;), I switched to the editor and didn't look away). `state == {}` is very very obvious.  `if bot.round == 1 and bot.turn == 0` is something that I have to think about.

Reply:
> One of our tests should do that, so it is clear how to do it.

I think we should be consistent here also. When we have a pattern that the students should employ for specific things, we should use it in all of the demos and tests. Otherwise it is very  likely that it will be overlooked. This is also one of the criticisms that I have heard.

Reply:
> > One of our tests should do that, so it is clear how to do it.
> 
> I think we should be consistent here also. When we have a pattern that the students should employ for specific things, we should use it in all of the demos and tests. Otherwise it is very likely that it will be overlooked. 
Sure, so where do you find a contradiction? I say that the reason not to use `if bot.round==1 and bot.turn==0` can not be "but then you have to create a state by hand when you want to test the move function for bot.turn==1". We should instead have at least one test that tests the second bot (I think we have none now), and that test should use the correct pattern. Testing for the second bot only makes sense for those move functions where the two bots have different behaviour. This is the case only for demo `one_and_one` I think. so that test should test both bots.
This is also one of the criticisms that I have heard.

Reply:
>  Otherwise it is very likely that it will be overlooked. This is also one of the criticisms that I have heard.

Well, in your presentation you did not even **mention** that we have tests for the demos and that both the tests and the demos are meant to be used as starting points for their own implementations. So, please, update the introduction, maybe go back to have slides if it helps you to remember what you want to say, but scrolling up-and-down on a webpage while mentioning random things that you can read there is **not** helping them to understand what all those files are there for.

Stuff is not overlooked so easily if you explicitly point people to it.

Reply:
>  `state == {}` is very very obvious. `if bot.round == 1 and bot.turn == 0` is something that I have to think about.

Yes, as I said, `state == {}` is obvious, but it does not make it clear **when** this is the case. When the students look at the demos they are still not familiar with the game, so they asked me at least in three different groups when does the `state` get initialized and why is the state sometimes `None`. I think all of this becomes clear with the `if bot.round == 1 and bot.turn == 0` condition.